api/client/client.go

```diff
@@ -81,21 +87,34 @@ func NewClient(cfg Config) (*Client, error) {
 		return conn, nil
 	})
 
+	// Attempt to connect to the server with each of the Credentials
+	// and keep the first successful connection.
+	var errs []error
 	var err error
-	if c.conn, err = grpc.Dial(constants.APIDomain,
-		dialer,
-		grpc.WithTransportCredentials(credentials.NewTLS(c.c.TLS)),
-		grpc.WithKeepaliveParams(keepalive.ClientParameters{
-			Time:                c.c.KeepAlivePeriod,
-			Timeout:             c.c.KeepAlivePeriod * time.Duration(c.c.KeepAliveCount),
-			PermitWithoutStream: true,
-		}),
-	); err != nil {
-		return nil, trail.FromGRPC(err)
-	}
+	for i, creds := range c.c.Credentials {
+		c.conn, err = grpc.Dial(constants.APIDomain,
+			dialer,
+			grpc.WithTransportCredentials(credentials.NewTLS(creds.TLS)),
+			grpc.WithKeepaliveParams(keepalive.ClientParameters{
+				Time:                c.c.KeepAlivePeriod,
+				Timeout:             c.c.KeepAlivePeriod * time.Duration(c.c.KeepAliveCount),
+				PermitWithoutStream: true,
+			}),
+		)
+		if err != nil {
+			// TODO (Joerger): if connecting to auth fails, try connecting via proxy
+			errs = append(errs, trace.Wrap(err, "Credentials[%v]", i))
+			continue
+		}
 
-	c.grpc = proto.NewAuthServiceClient(c.conn)
-	return c, nil
+		c.grpc = proto.NewAuthServiceClient(c.conn)
+		// TODO (Joeger): we can't test the authentication of the connection without pinging the server,
+		// but doing so here would break the current uses of Client (tests). This Credentials logic could
+		// instead be used to fallback to other credentials whenever a grpc request returns a connection error.
```

Comment:
How would the tests break?

Calling `Ping` is part of the RFD, will tell us if the client can talk with the server plus it will tell us if the credentials were valid.

From the RFD 10:

> Client should verify that its implementation is compatible with Teleport's server version by calling Ping and checking the version similar to tsh logic.

Reply:
Some of the tests that use `lib/client` do not provide valid TLS configs or connections - see this failed build https://drone.gravitational.io/gravitational/teleport/6354/1/6.

I think I can get around this issue by making a bare bones `NewTLSClient` function in `api/client` for `lib/client` to use, rather than the main `client.New` function which uses credential providers and everything.